src/http/endpoints/update_profile.rs

```diff
@@ -0,0 +1,86 @@
+use crate::ayb_db::db_interfaces::AybDb;
+use crate::ayb_db::models::{Entity, InstantiatedEntity, Link};
+use crate::error::AybError;
+use crate::http::structs::EntityPath;
+use crate::http::utils::unwrap_authenticated_entity;
+use crate::http::verification::is_verified;
+use crate::http::web_frontend::WebFrontendDetails;
+use actix_web::{post, web, HttpResponse};
+use serde::Deserialize;
+use std::str::FromStr;
+use url::Url;
+
+#[derive(Deserialize)]
+struct ProfileLinkUpdate {
+    pub url: String,
+}
+
+#[derive(Deserialize)]
+struct ProfileUpdate {
+    pub display_name: Option<String>,
+    pub description: Option<String>,
+    pub workplace: Option<String>,
+    pub links: Vec<ProfileLinkUpdate>,
+}
+
+#[post("/v1/entity/{entity}")]
+pub async fn update_profile(
+    path: web::Path<EntityPath>,
+    profile: web::Json<ProfileUpdate>,
+    ayb_db: web::Data<Box<dyn AybDb>>,
+    web_info: web::Data<Option<WebFrontendDetails>>,
+    authenticated_entity: Option<web::ReqData<InstantiatedEntity>>,
+) -> Result<HttpResponse, AybError> {
+    let authenticated_entity = unwrap_authenticated_entity(&authenticated_entity)?;
+    let entity_slug = &path.entity;
+    let profile = profile.into_inner();
+
+    if entity_slug != &authenticated_entity.slug {
+        return Ok(HttpResponse::Unauthorized().finish());
+    }
+
+    let instantiated_entity = ayb_db.get_entity_by_slug(entity_slug).await?;
+    let links = if let Some(web_info) = Option::as_ref(&**web_info) {
+        let mut links = vec![];
+        for link in profile.links.into_iter() {
+            let url = Url::parse(&link.url)?;
+            links.push(Link {
+                url: link.url,
+                verified: is_verified(
+                    url,
+                    Url::from_str(&web_info.profile(entity_slug))
+                        .expect("invalid web frontend url"),
```

Comment:
Is there a way to avoid a panic here?

Reply:
This panic is similar to this one: https://github.com/marcua/ayb/blob/7bcd087028b86e1dcce4b552263ff6a5ba186293/src/http/web_frontend.rs#L28

It only can happen when the web frontend didn't send the correct endpoints, which shouldn't happen ever. And if it does happen, the instance administrator must know that something is off.

From the Rust book:
> It’s advisable to have your code panic when it’s possible that your code could end up in a bad state. In this context, a bad state is when some assumption, guarantee, contract, or invariant has been broken, such as when invalid values, contradictory values, or missing values are passed to your code.

In that case, the web frontend has sent invalid values and broken the guarantee that all endpoints are valid URLs.

Reply:
Great explanation, thank you!